Mathlib/Algebra/MonoidAlgebra/NoZeroDivisors.lean

```diff
@@ -100,4 +130,126 @@ theorem mul_apply_add_eq_mul_of_uniqueAdd [Add A] {f g : R[A]} {a0 b0 : A}
+It is proved with a further `CovariantClass` assumption. -/
+theorem single_mul_apply_of_le {a t : A} (r : R) {f : AddMonoidAlgebra R A}
+    (ft : ∀ a ∈ f.support, a ≤ t) :
+    (AddMonoidAlgebra.single a r * f) (a + t) = r * f t := by
+  rw [AddMonoidAlgebra.single_mul_apply_aux]
+  intro t' ht'
+  specialize ft t' ht'
+  refine ⟨fun h => ?_, fun h => h ▸ rfl⟩
+  contrapose h
+  exact add_lt_add_left (ft.lt_of_ne h) a |>.ne
+
+/--  The "bottom" element of `Finsupp.single a r * f` is the product of `r` and
+the "bottom" element of `f`.  Here, "bottom" is simply a lower bound for the elements
+of the support of `f` (e.g. the product is `0` if "bottom" is not a minimum).
+The corresponding statement for a general product `f * g` is `AddMonoidAlgebra.mul_apply_of_le'`.
+It is proved with a further `CovariantClass` assumption. -/
+theorem single_mul_apply_of_le' {a b : A} {f : AddMonoidAlgebra R A} (r : R)
+    (fb : ∀ a ∈ f.support, b ≤ a) :
+    (AddMonoidAlgebra.single a r * f) (a + b) = r * f b :=
+  single_mul_apply_of_le (A := Aᵒᵈ) _ fb
+
+end Left
+
+section Right
+variable [AddRightStrictMono A]
+
+/--  The "top" element of `f * AddMonoidAlgebra.single a r` is the product of
+the "top" element of `f` and `r`.  Here, "top" is simply an upper bound for the elements
+of the support of `f` (e.g. the product is `0` if "top" is not a maximum).
+The corresponding statement for a general product `f * g` is `AddMonoidAlgebra.mul_apply_of_le`.
+It is proved with a further `CovariantClass` assumption. -/
+theorem mul_single_apply_of_le {a t : A} (r : R) {f : AddMonoidAlgebra R A}
+    (ft : ∀ a ∈ f.support, a ≤ t) :
+    (f * AddMonoidAlgebra.single a r) (t + a) = f t * r := by
+  rw [AddMonoidAlgebra.mul_single_apply_aux]
+  intro t' ht'
+  specialize ft t' ht'
+  refine ⟨fun h => ?_, fun h => h ▸ rfl⟩
+  contrapose h
+  exact add_lt_add_right (ft.lt_of_ne h) a |>.ne
+
+/--  The "bottom" element of `Finsupp.single a r * f` is the product of `r` and
+the "bottom" element of `f`.  Here, "bottom" is simply a lower bound for the elements
+of the support of `f` (e.g. the product is `0` if "bottom" is not a minimum).
+The corresponding statement for a general product `f * g` is `AddMonoidAlgebra.mul_apply_of_le'`.
+It is proved with a further `CovariantClass` assumption. -/
+theorem mul_single_apply_of_le' {a b : A} {f : AddMonoidAlgebra R A} (r : R)
+    (fb : ∀ a ∈ f.support, b ≤ a) :
+    (f * AddMonoidAlgebra.single a r) (b + a) = f b * r :=
+  mul_single_apply_of_le (A := Aᵒᵈ) _ fb
+
+end Right
+
+section Both
+
+variable [AddLeftStrictMono A] [AddRightStrictMono A]
+variable {a t b : A} {f g : AddMonoidAlgebra R A}
+
+/--  The "top" element of `f * g` is the product of the "top" elements of `f` and of `g`.
+Here, "top" is simply an upper bound for the elements of the support of the corresponding
+polynomial (e.g. the product is `0` if "top" is not a maximum). -/
+theorem mul_apply_of_le (fa : ∀ i ∈ f.support, i ≤ a) (gt : ∀ i ∈ g.support, i ≤ t) :
```

Comment:
What's the replacement for this lemma?

Reply:
Hmm, the closest that I'm finding is [AddMonoidAlgebra.leadingCoeff_mul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/MonoidAlgebra/Degree.html#AddMonoidAlgebra.leadingCoeff_mul), but I agree that `leadingCoeff` is not as flexible as what this lemma does.

Maybe this lemma should be extracted as well.